sdk/python/feast/diff/FcoDiff.py

```diff
@@ -80,23 +83,138 @@ def tag_proto_objects_for_keep_delete_add(
     )
+
+
+def extract_objects_for_keep_delete_update_add(
+    registry: Registry, current_project: str, desired_repo_contents: RepoContents,
+) -> Tuple[
+    Dict[str, Set[Fco]], Dict[str, Set[Fco]], Dict[str, Set[Fco]], Dict[str, Set[Fco]]
+]:
+    """
+    Returns the objects in the registry that must be modified to achieve the desired repo state.
+
+    Args:
+        registry: The registry storing the current repo state.
+        current_project: The Feast project whose objects should be compared.
+        desired_repo_contents: The desired repo state.
+    """
+    objs_to_keep = {}
+    objs_to_delete = {}
+    objs_to_update = {}
+    objs_to_add = {}
+
+    registry_object_type_to_objects: Dict[str, List[Any]]
+    registry_object_type_to_objects = {
+        "entities": registry.list_entities(project=current_project),
+        "feature_views": registry.list_feature_views(project=current_project),
+        "on_demand_feature_views": registry.list_on_demand_feature_views(
+            project=current_project
+        ),
+        "request_feature_views": registry.list_request_feature_views(
+            project=current_project
+        ),
+        "feature_services": registry.list_feature_services(project=current_project),
+    }
+
+    for object_type in REGISTRY_OBJECT_TYPES:
+        (
+            to_keep,
+            to_delete,
+            to_update,
+            to_add,
+        ) = tag_objects_for_keep_delete_update_add(
+            registry_object_type_to_objects[object_type],
+            getattr(desired_repo_contents, object_type),
+        )
+
+        objs_to_keep[object_type] = to_keep
+        objs_to_delete[object_type] = to_delete
+        objs_to_update[object_type] = to_update
+        objs_to_add[object_type] = to_add
+
+    return objs_to_keep, objs_to_delete, objs_to_update, objs_to_add
+
+
+def diff_between(
+    registry: Registry, current_project: str, desired_repo_contents: RepoContents,
+) -> RegistryDiff:
+    """
+    Returns the difference between the current and desired repo states.
+
+    Args:
+        registry: The registry storing the current repo state.
+        current_project: The Feast project for which the diff is being computed.
+        desired_repo_contents: The desired repo state.
+    """
+    diff = RegistryDiff()
+
+    (
+        objs_to_keep,
+        objs_to_delete,
+        objs_to_update,
+        objs_to_add,
+    ) = extract_objects_for_keep_delete_update_add(
+        registry, current_project, desired_repo_contents
+    )
+
+    for object_type in REGISTRY_OBJECT_TYPES:
+        objects_to_keep = objs_to_keep[object_type]
+        objects_to_delete = objs_to_delete[object_type]
+        objects_to_update = objs_to_update[object_type]
+        objects_to_add = objs_to_add[object_type]
+
+        for e in objects_to_add:
+            diff.add_fco_diff(
+                FcoDiff(
+                    e.name,
```

Comment:
nit: mind adding the named parameters? e.g. name=e.name. Will make this easier to read